src/site/content/en/notifications/codelab-notifications-service-worker/index.md

```diff
@@ -0,0 +1,201 @@
+  // (so you can tell the notifications apart during testing!)
+  let randy = Math.floor(Math.random() * 100);
+  let notification = {
+    title: 'Test ' + randy,
+    options: { body: 'Test body ' + randy }
+  };
+  // Get a reference to the service worker registration.
+  let registration = await getRegistration();
+  // Check that the service worker registration exists.
+  if (registration) {
+    // Check that a service worker controller exists before
+    // trying to access the postMessage method.
+    if (navigator.serviceWorker.controller) {
+      navigator.serviceWorker.controller.postMessage(notification);
+    } else {
+      console.log('No service worker controller found. Try a soft reload.');
+    }
+  }
+}
+```
+
+Here's what that code is doing:
+
+*   `sendNotification` is an asynchronous function, so you can use `await` to get a reference to the service worker registration.
+
+*   The service worker's `postMessage` method sends data from the app to the service worker. See the [MDN documentation on  postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Worker/postMessage) for more information.
+
+*   The code checks for the presence of the `navigator.serviceWorker.controller` property before trying to access the `postMessage` function. `navigator.serviceWorker.controller` will be `null` if there is no active service worker, or if the page has been force refreshed (`Shift+`**Reload**). See the [ServiceWorker controller documentation on MDN](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerContainer/controller) for more information.
+
+## Handle notifications in the service worker
+
+In this step, you'll write code in the service worker that will handle messages posted to it and display notifications to the user.
+
+In the embedded Glitch editor, open `public/service-worker.js`. Add the following code to the end of the file:
+
+```js
+// Show notification when received
+self.addEventListener('message', (event) => {
+  let notification = event.data;
+  self.registration.showNotification(
+    notification.title,
+    notification.options
+  ).catch((error) => {
+    console.log(error);
+  });
+});
+```
+
+Here's a quick explanation:
+
+*   `self` is a reference to the service worker itself.
+
+*   While the service worker now handles displaying notifications, the main app UI is still responsible for getting notification permission from the user. If permission is not granted, the promise returned by `showNotification` is rejected. The code above uses a `catch` block to avoid an uncaught `Promise` rejection error and handle this error a little more gracefully.
+
+If you got stuck, see [glitch.com/edit/#!/codelab-notifications-service-worker-completed](https://glitch.com/edit/#!/codelab-notifications-service-worker-completed) for the completed code.
+
+Go on to the next codelab in this series: [Build a push notifications server](/codelab-notifications-push-server).
```

Comment:
We usually don't string together codelabs, preferring instead to string together the guides that codelabs are associated with.

We have plans to create a multi-step codelab experience after CDS and perhaps it would make more sense to string things together then? @kaycebasques @mfriesenhahn for their thoughts.

Reply:
This seemed like a reasonable compromise to me given the lack of multi-step codelabs. Defer to you and Kayce.

Reply:
My concern is if we start creating a new UX experience that is unlike anything else on the site. I prefer being consistent, even when it's a bit annoying, because it matches user's prior experience.

Reply:
Can we talk this one through? I'm normally all about predictable UX, but I'm honestly unsure how disruptive this will be. And writing duplicative posts just so we can have something to associate a codelab with seems un-good. (Though I may be misunderstanding what you're suggesting.) And the short-term strategy might depend on timeline for multi-step codelabs?

Reply:
yep, pinged you on chat

Reply:
I'll think about it while I edit the collection